proposals/target-typed-new.md

```diff
@@ -31,37 +31,29 @@ Instantiate an object without spelling out the type.
 private readonly static object s_syncObj = new();
 ```
 
-## Detailed design
+## Specification
 [design]: #detailed-design
 
-The *object_creation_expression* syntax would be modified to make the *type* optional when parentheses are present. This is required to address the ambiguity with *anonymous_object_creation_expression*.
+A new syntactic form, *target_typed_new* of the *object_creation_expression* is accepted in which the *type* is optional.
+
 ```antlr
 object_creation_expression
-    : 'new' type? '(' argument_list? ')' object_or_collection_initializer?
+    : 'new' type '(' argument_list? ')' object_or_collection_initializer?
     | 'new' type object_or_collection_initializer
+    | target_typed_new
+    ;
+target_typed_new
+    : 'new' '(' argument_list? ')' object_or_collection_initializer?
     ;
 ```
 
-A target-typed `new` is convertible to any type. As a result, it does not contribute to overload resolution. This is mainly to avoid unpredictable breaking changes.
-
-The argument list and the initializer expressions will be bound after the type is determined.
-
-The type of the expression would be inferred from the target-type which would be required to be one of the following:
-
-- **Any struct type** (including tuple types)
-- **Any reference type** (including delegate types)
-- **Any type parameter** with a constructor or a `struct` constraint
+A *target_typed_new* expression does not have a type. However, there is a new *object creation conversion* that is an implicit conversion from expression, that exists from a *target_typed_new* to every type. It is an error for this conversion to be applied unless the type being converted to is
+- a `struct` type; or
+- a `class` type; or
+- a `delegate` type; or
+- a type parameter.
```

Comment:
I think this is missing interfaces, due to COM.

Reply:
Ah, I see the deleted diff has an explicit exception. I'm not sure that I agree with the exception, it feels inconsistent for no reason.

Reply:
Interface were excluded because we only target *constructible types* as per LDM. Interfaces were not deemed to be constructible.

Reply:
Interface were excluded because we only target constructible types as per LDM. Interfaces were not deemed to be constructible.

Reply:
Will bring this up today. Personally, I think that's true everywhere except the COM case, and limiting it there just feels inconsistent.

Reply:
The existing language spec does not permit an object creation expression to specify an interface type. See [https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#object-creation-expressions](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#object-creation-expressions)
This is not the place to start doing that.

---
In reply to: [397967636](https://github.com/dotnet/csharplang/pull/3310#discussion_r397967636) [](ancestors = 397967636)